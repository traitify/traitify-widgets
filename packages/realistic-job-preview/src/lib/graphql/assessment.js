export const create = `
  mutation createAssessment(
    $localeKey: String!,
    $profileID: String!,
    $surveyID: String!
  ) {
    createAssessment(
      localeKey: $localeKey,
      profileId: $profileID,
      surveyId: $surveyID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const get = `
  query assessment(
    $assessmentID: ID!
  ) {
    assessment(
      id: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const path = "/realistic-job-previews/graphql";

export const start = `
  mutation startAssessment(
    $assessmentID: ID!
  ) {
    startAssessment(
      id: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;

export const update = `
  mutation updateAssessmentAnswer(
    $answers: [AssessmentAnswer],
    $assessmentID: String!
  ) {
    updateAssessmentAnswer(
      answers: $answers,
      assessmentId: $assessmentID
    ) {
      completedAt
      failedReason
      id
      insertedAt
      instructions
      isFit
      jobId
      localeKey
      profileId
      responses {
        id
        questionId
        questionText
        responseOptions {
          responseOptionId
          responseOptionText
        }
        selectedResponseOptionId
        selectedResponseOptionText
      }
      startedAt
      status
      surveyId
      totalCorrectResponses
      updatedAt
      videos {
        thumbnailUrl
        videoUrl
      }
    }
  }
`;
