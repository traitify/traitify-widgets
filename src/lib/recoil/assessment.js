import {selector, selectorFamily} from "recoil";
import {assessmentsState} from "./assessments";
import {
  activeState,
  activeIDState,
  activeTypeState,
  cacheState,
  graphqlState,
  httpState,
  localeState,
  safeCacheKeyState
} from "./base";

// TODO: Return error instead of null for cognitive and external
// return {errors: response.errors, id};
export const cognitiveAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const params = {
      query: GraphQL.cognitive.get,
      variables: {localeKey: get(localeState), testID: id}
    };
    const response = await http.post({path: GraphQL.cognitive.path, params});
    if(response.errors) {
      console.warn("cognitive-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.cognitiveTest;
    if(!assessment?.completed) { return assessment; }

    cache.set(cacheKey, assessment);

    return assessment;
  },
  key: "assessment/cognitive"
});

export const externalAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const query = {
      params: {
        query: GraphQL.external.get,
        variables: {id}
      },
      path: GraphQL.external.path
    };
    if(http.version === "v1") { query.version = GraphQL.external.version; }

    const response = await http.post(query);
    if(response.errors) {
      console.warn("external-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.getAssessment;
    if(!assessment?.completedAt) { return assessment; }

    cache.set(cacheKey, assessment);

    return assessment;
  },
  key: "assessment/external"
});

export const personalityAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached) { return cached; }

    const params = {
      data: "archetype,blend,instructions,recommendation,slides,types,traits",
      locale_key: get(localeState)
    };
    const http = get(httpState);
    const response = await http.get({path: `/assessments/${id}`, params});
    if(!response?.completed_at) { return response; }

    cache.set(cacheKey, response);

    return response;
  },
  key: "assessment/personality"
});

export const genericAssessmentQuery = selectorFamily({
  get: (id) => async({get}) => {
    if(!id) { return null; }

    const cache = get(cacheState);
    const cacheKey = get(safeCacheKeyState({id, type: "assessment"}));
    const cached = cache.get(cacheKey);
    if(cached && !cached.completed) { return cached; }

    const GraphQL = get(graphqlState);
    const http = get(httpState);
    const params = {
      query: GraphQL.generic.questions,
      variables: {assessmentID: id}
    };

    const response = await http.post({path: GraphQL.generic.path, params});
    if(response.errors) {
      console.warn("generic-assessment", response.errors); /* eslint-disable-line no-console */
      return null;
    }

    const assessment = response.data.genericSurveyQuestions;
    if(!assessment?.completedAt) { return assessment; }

    const query = GraphQL.generic.result;
    const variables = {assessmentID: id};

    const resultResponse = await http.post(GraphQL.generic.path, {query, variables});
    if(resultResponse.errors) { return null; }
    const result = resultResponse.data.genericAssessmentResult.assessment;

    cache.set(cacheKey, result);

    return result;
  },
  key: "assessment/generic"
});

export const assessmentQuery = selectorFamily({
  get: ({id, surveyType}) => async({get}) => {
    if(surveyType === "cognitive") { return get(cognitiveAssessmentQuery(id)); }
    if(surveyType === "external") { return get(externalAssessmentQuery(id)); }
    if(surveyType === "generic") { return get(genericAssessmentQuery(id)); }
    if(surveyType === "personality") { return get(personalityAssessmentQuery(id)); }

    return null;
  },
  key: "assessment"
});

export const activeAssessmentQuery = selector({
  get: async({get}) => {
    const id = get(activeIDState);
    const surveyType = get(activeTypeState);

    return get(assessmentQuery({id, surveyType}));
  },
  key: "assessment/active"
});

export const completedAssessmentQuery = selectorFamily({
  get: ({surveyType}) => async({get}) => {
    if(!surveyType) {
      const active = get(activeState);
      if(!active?.completed) { return null; }

      return get(activeAssessmentQuery);
    }

    const assessments = get(assessmentsState);
    if(!assessments) { return null; }

    const assessment = assessments
      .find(({completed, surveyType: type}) => type === surveyType && completed);
    if(!assessment) { return null; }

    return get(assessmentQuery({id: assessment.id, surveyType}));
  },
  key: "assessment/completed"
});
