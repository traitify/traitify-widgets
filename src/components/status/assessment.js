import {faCheck} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
import {useSetRecoilState} from "recoil";
import Icon from "components/common/icon";
import get from "lib/common/object/get";
import useListener from "lib/hooks/use-listener";
import useOption from "lib/hooks/use-option";
import {activeState} from "lib/recoil";
import style from "./style.scss";

// TODO: Extract text to translate
const translations = {
  complete: "Complete",
  loading: "Loading", // Already translated
  status: {
    start: "Start Assessment"
  },
  survey: {
    cognitive_assessment: "Cognitive Assessment",
    external_assessment: "External Assessment",
    personality_assessment: "Personality Assessment"
  }
};

function Button({assessment}) {
  const listener = useListener();
  const options = useOption("status") || {};
  const redirect = get(options, "allowRedirect", true);
  const setActive = useSetRecoilState(activeState);

  if(assessment.completed) {
    return <button disabled={true} type="button">{translations.complete}</button>;
  }

  if(assessment.skipped) {
    return <button disabled={true} type="button">Skipped</button>;
  }

  if(assessment.link && redirect) {
    return <a href={assessment.link}>{translations.status.start}</a>;
  }

  if(assessment.loading) {
    return <button disabled={true} type="button">{translations.loading}</button>;
  }

  const start = () => {
    listener.trigger("Survey.start", {assessment});
    setActive({...assessment});
  };

  return <button onClick={start} type="button">{translations.status.start}</button>;
}

Button.propTypes = {
  assessment: PropTypes.shape({
    completed: PropTypes.bool,
    link: PropTypes.string,
    loading: PropTypes.bool,
    skipped: PropTypes.bool
  }).isRequired
};

function Assessment({assessment}) {
  const surveyName = assessment.surveyName || translations.survey[`${assessment.surveyType}_assessment`];

  return (
    <div className={[style.assessment, assessment.completed && style.inactive].filter(Boolean).join(" ")}>
      <div>
        {assessment.completed && <Icon className={style.icon} icon={faCheck} />}
        <span className={style.text}>{surveyName}</span>
      </div>
      <Button assessment={assessment} />
    </div>
  );
}

Assessment.propTypes = {
  assessment: PropTypes.shape({
    completed: PropTypes.bool,
    skipped: PropTypes.bool,
    surveyName: PropTypes.string,
    surveyType: PropTypes.string.isRequired
  }).isRequired
};

export default Assessment;
