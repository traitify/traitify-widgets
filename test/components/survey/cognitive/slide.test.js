/** @jest-environment jsdom */
import Component from "components/survey/cognitive/slide";
import ComponentHandler from "support/component-handler";
import assessment from "support/json/assessment/cognitive.json";
import useWindowMock from "support/hooks/use-window-mock";

const defaultQuestions = assessment.questions.slice(1, 3);
const loadedQuestions = defaultQuestions.map((question) => ({...question, loaded: true}));

describe("Slide", () => {
  let originalDate;
  let props;

  beforeAll(() => {
    originalDate = Date.now;
  });

  beforeEach(() => {
    const now = Date.now();
    Date.now = jest.fn().mockName("now").mockReturnValue(now);

    props = {
      onSelect: jest.fn().mockName("onSelect"),
      question: loadedQuestions[0],
      translate: jest.fn().mockName("translate").mockImplementation((value) => value)
    };
  });

  afterAll(() => {
    Date.now = originalDate;
  });

  describe("setup", () => {
    let component;

    beforeEach(() => {
      component = new ComponentHandler(<Component {...props} />);
    });

    it("disables confirm button", () => {
      const button = component.instance.findAllByType("button")
        .find((element) => element.children[0] === "cognitive_confirm_button");

      expect(button.props.disabled).toBe(true);
      expect(button.props.onClick).toBeNull();
    });

    it("sets start time", () => {
      expect(Date.now).toHaveBeenCalled();
    });
  });

  describe("skip", () => {
    useWindowMock("confirm");

    it("calls select", () => {
      window.confirm.mockReturnValue(true);
      const component = new ComponentHandler(<Component {...props} />);
      component.act(() => component.instance.findAllByProps({className: "choice"})[1].children[0].props.onClick());
      Date.now.mockReturnValue(Date.now() + 2000);
      component.act(() => component.findByText("cognitive_skip_button").props.onClick());

      expect(window.confirm).toHaveBeenCalledTimes(1);
      expect(props.onSelect).toHaveBeenCalledWith({skipped: true, timeTaken: 2000});
    });

    it("calls skip if present", () => {
      props.onSkip = jest.fn().mockName("onSkip");
      const component = new ComponentHandler(<Component {...props} />);
      component.act(() => component.instance.findAllByProps({className: "choice"})[1].children[0].props.onClick());
      Date.now.mockReturnValue(Date.now() + 2000);
      component.act(() => component.findByText("cognitive_skip_button").props.onClick());

      expect(window.confirm).not.toHaveBeenCalled();
      expect(props.onSelect).not.toHaveBeenCalled();
      expect(props.onSkip).toHaveBeenCalledWith({timeTaken: 2000});
    });

    it("does nothing if not confirmed", () => {
      window.confirm.mockReturnValue(false);
      const component = new ComponentHandler(<Component {...props} />);
      component.act(() => component.instance.findAllByProps({className: "choice"})[1].children[0].props.onClick());
      Date.now.mockReturnValue(Date.now() + 2000);
      component.act(() => component.findByText("cognitive_skip_button").props.onClick());

      expect(window.confirm).toHaveBeenCalledTimes(1);
      expect(props.onSelect).not.toHaveBeenCalled();
    });
  });

  describe("update", () => {
    let button;
    let component;

    beforeEach(() => {
      component = new ComponentHandler(<Component {...props} />);
      props.onSelect.mockImplementation(() => (
        component.updateProps({question: loadedQuestions[1]})
      ));
      component.act(() => component.instance.findAllByProps({className: "choice"})[0].children[0].props.onClick());
      Date.now.mockReturnValue(Date.now() + 2000);
      Date.now.mockClear();
      component.act(() => component.findByText("cognitive_confirm_button").props.onClick());
      button = component.findByText("cognitive_confirm_button");
    });

    it("disables confirm button", () => {
      expect(button.props.disabled).toBe(true);
      expect(button.props.onClick).toBeNull();
    });

    it("resets start time", () => {
      expect(Date.now).toHaveBeenCalled();
    });
  });

  it("renders component", () => {
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("renders loading", () => {
    props.question = defaultQuestions[0];
    const component = new ComponentHandler(<Component {...props} />);

    expect(component.tree).toMatchSnapshot();
  });

  it("submits selected answer", () => {
    const component = new ComponentHandler(<Component {...props} />);
    component.act(() => (
      component.instance.findAllByProps({className: "choice"})[1].children[0].props.onClick()
    ));
    Date.now.mockReturnValue(Date.now() + 2000);
    component.act(() => component.findByText("cognitive_confirm_button").props.onClick());

    expect(props.onSelect).toHaveBeenCalledWith({
      answerId: props.question.responses[1].id,
      timeTaken: 2000
    });
  });
});
