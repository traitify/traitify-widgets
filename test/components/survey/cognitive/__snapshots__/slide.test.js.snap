// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide renders component 1`] = `
<div
  className=""
>
  <div
    className="question"
  >
    <img
      alt="cognitive_question_alt_text"
      src="https://cdn.traitify.com/images/cognitive_tests/10017/q.png"
    />
  </div>
  <div
    className="choices"
  >
    <div
      className="choicesContainer"
    >
      <div
        className="choice"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <img
            alt="cognitive_response_alt_text"
            className={null}
            src="https://cdn.traitify.com/images/cognitive_tests/10017/a.png"
          />
        </button>
      </div>
      <div
        className="choice"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <img
            alt="cognitive_response_alt_text"
            className={null}
            src="https://cdn.traitify.com/images/cognitive_tests/10017/o1.png"
          />
        </button>
      </div>
      <div
        className="choice"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <img
            alt="cognitive_response_alt_text"
            className={null}
            src="https://cdn.traitify.com/images/cognitive_tests/10017/o3.png"
          />
        </button>
      </div>
      <div
        className="choice"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <img
            alt="cognitive_response_alt_text"
            className={null}
            src="https://cdn.traitify.com/images/cognitive_tests/10017/o2.png"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    className="buttons"
  >
    <button
      className="btnSkip"
      onClick={[Function]}
      type="button"
    >
      cognitive_skip_button
    </button>
    <button
      className="btnDisabled"
      disabled={true}
      onClick={null}
      type="button"
    >
      cognitive_confirm_button
    </button>
  </div>
</div>
`;

exports[`Slide renders loading 1`] = `
<div
  className="loadingContainer"
>
  <div
    className="loading"
  >
    <i />
    <i />
  </div>
</div>
`;
